design/authenticators/authn_jwt/authn_jwt_solution_design.md

```diff
@@ -0,0 +1,465 @@
+implementations which based on JWT.
+
+Each authentication implementation will be represented with a different JWT configuration, for example:
+1. VendorX - authenticate with JWT which signed by static key
+2. VendorY - authenticate with JWT and have the identity (user/host) in some format under the audience claim
+3. VendorZ - authenticate with JWT and we must validate some custom claims (projectID, location)
+
+### Design
+The following class diagram, represent the responsibilities and the boundaries of the JWT generic authenticator.
+#### Class diagram
+![Authn jwt class diagram](authn-jwt-class-diagram.png)
+
+According to the above diagram we will have the flexibility to support any vendor JWT 
+implementation no matter what are the answers of the following questions:
+1. From where to fetch the public key token?
+   
+   1.1 Provider URI ? like we do in Azure, OIDC and GCP
+   
+   1.2 Special JWKS url ? Like Gitlab implementation
+
+   1.3 Static key? 
+
+2. Which JWT claim to validate ? exp, nbf, iss, aud, jti, iat, sub
+   * Note: All of them are optional, but we already saw differences between some vendors implementations
+    ![Authn jwt vendors claims comparison](authn-jwt-vendors-claims-comparison.png)
+
+3. Where is the identity to authorized ?
+    
+    3.1. In the URL request? like GCP and Azure
+   
+    3.2. Inside the token? Under which claim? Like OIDC
+    
+    Note:
+    In 1st phase we will support both 3.1, 3.2
+
+4. How to validate the user restrictions? 
+    
+    4.1. Where are those restrictions ? in all of our authenticators the restrictions reside under the 
+   identity (user/host) policy, but according to this design we have the
+   flexibility to support other approaches in the future.
+   
+    4.2. What are the constraints of the restrictions? mandatory? optional? permitted ? at least one? OR relation? AND relation? 
+   
+    4.3. Validate restrictions value against the token?
+   
+    * Comparison logic? ==? contains?
+    
+    * Key mapping logic? for example: auth-gcp/project-id is subject claim in JWT
+
+#### Sequence diagram
+![Authn jwt flow](authn-jwt-flow.png)
+
+#### Business logic
+
+##### Fetch identity
+For 1st milestone we will support 2 use cases:
+1. User/Host identity which delivered in the Rest request parameter. for example:
+   user-test is the identity in the following request. 
+   This parameter is optional
+
+   
+        POST https://<Conjur-server-hostname>/authn-jwt/<service-id>/<account>/user-test/authenticate
+2. User/Host identity which reside inside the token under claim key which configured in `jwt-user-property`
+   in the authenticator policy. for example: if the identity reside in `user_email` claim we will load the 
+   following policy:
+   ```yaml
+    - !policy
+    id: conjur/authn-jwt/VendorY
+    body:
+    - !webservice
+    
+    - !variable
+      id: jwt-user-property #TBD on the property name
```

Comment:
jwt-user-property should be id-token-app-property
should replace everywhere

Reply:
why `is-token` and  not `jwt` ? what if we will have a use case of JWT token which contains the user but it is not id token ?

Reply:
I want it to be generic, maybe token-app-property without id, but I don't want it to be specific for jwt so we can re-use this variable later if we'll need it.